diagnostic_feedback/quiz.py

```diff
@@ -0,0 +1,436 @@
+        """
+        ajax handler to save data after applying required validation & filtration
+        :param data: step data to save
+        :param suffix:
+        :return: response dict
+        """
+
+        success = True
+        response_message = ""
+        step = data.get('step', '')
+
+        if not step:
+            success = False
+            response_message = self._('missing step number')
+        else:
+            try:
+                is_valid_data, response_message = Validator.validate(self, data)
+                if is_valid_data:
+                    response_message = MainHelper.save_filtered_data(self, data)
+                else:
+                    success = False
+
+            except Exception as ex:
+                success = False
+                response_message += ex.message if ex.message else str(ex)
+
+        return {'step': step, 'success': success, 'msg': response_message}
+
+    @XBlock.json_handler
+    def save_choice(self, data, suffix=''):
+        """
+        save student choice for a question after validations
+        :param data: answer data
+        :param suffix:
+        :return: response dict
+        """
+
+        student_result = ""
+        response_message = ""
+
+        try:
+            success, response_message = Validator.validate_student_answer(self, data)
+            if success:
+                question_id = data['question_id']
+
+                # save student answer
+                self.student_choices[question_id] = data['student_choice']
+                if self.current_step < data['currentStep']:
+                    self.current_step = data['currentStep']
+
+                # calculate feedback result if user answering last question
+                if data['isLast']:
+                    student_result = self.get_result()
+
+                    if not self.completed:
+                        # Save the latest score and make quiz completed
+                        self.runtime.publish(self, 'grade', {
+                            'value': 1.0,
+                            'max_value': 1.0
+                        })
+                        self.completed = True
+
+                    if my_api:
+                        log.info("have sub_api instance")
+                        # Also send to the submissions API:
+                        item_key = self.student_item_key
+                        item_key['item_id'] = self.get_block_id()
+                        submission_data = self.create_submission_data()
+                        submission_data['final_result'] = student_result
+                        my_api.create_submission(item_key, json.dumps(submission_data))
+
+                response_message = self._("Your response is saved")
+        except Exception as ex:
+            success = False
+            response_message += str(ex)
+        return {'success': success, 'student_result': student_result, 'response_msg': response_message}
```

Comment:
Above, in `save_data()`, the response message is stored in `'msg'`.  Here it's `'response_msg'`.  Unless there's a reason for the discrepancy, keep it consistent.